File: ruby/test/aoc_toolbox_test.rb
require 'test/unit'

require_relative '../aoc_toolbox.rb'

class AocToolboxTest < Test::Unit::TestCase

    def test_lines
        aoc = InputFile.new
        assert(aoc.lines == ["1337", "42"] )
    end

    def test_lines_ints
        aoc = InputFile.new
        assert(aoc.lines_ints == [1337, 42])
    end

    def test_fields_regexp
        aoc = InputFile.new(filename: "input-fields.txt")
        assert(aoc.linefields_regexp(/(\w+): (\d+)\s(\d+)\s(\d+)/)[0] == ['kalle', '123','456','879'])
    end

    def test_sections
        aoc = InputFile.new(filename: "input-sections.txt")
        assert(aoc.sections[3].lines == ['section 4', 'banana'])
    end

    def test_matrix_char
        aoc = InputFile.new(filename: "input-matrix.txt")
        assert(aoc.matrix_char(1, 4) == '.')
        assert(aoc.matrix_char(3, 15) == '#')
    end

    def test_sub_matrix
        aoc = InputFile.new(filename: "input-matrix.txt")
        assert(aoc.sub_matrix(1,1,2,2).lines == ['.#','..'])
    end
end

File: javascript/test/aoc_toolbox.test.js
const { InputData } = require('../aoc_toolbox')

test('lines()', () => {
    const i = new InputData()
    expect(i.lines()[0]).toBe('1337')
    expect(i.lines()[1]).toBe('42')
})

test('linesInts()', () => {
    const i = new InputData()
    expect(i.linesInts()[0]).toBe(1337)
    expect(i.linesInts()[1]).toBe(42)
})

test('linefieldsRegexp()', () => {
    const i = new InputData({ filename: 'input-fields.txt' })
    const r = /(\w+): (\d+)\s(\d+)\s(\d+)/
    const fieldArray = i.linefieldsRegexp(r)
    expect(fieldArray[0][0]).toBe('kalle')
    expect(fieldArray[0][1]).toBe('123')
    expect(fieldArray[0][2]).toBe('456')
    expect(fieldArray[0][3]).toBe('879')
})

test('sections()', () => {
    const i = new InputData({ filename: 'input-sections.txt' })
    expect(i.sections()[3].lines()[1]).toBe('banana')
})

test('matrixChar()', () => {
    const i = new InputData({ filename: 'input-matrix.txt' })
    expect(i.matrixChar(1, 4)).toBe('.')
    expect(i.matrixChar(3, 15)).toBe('#')
})

test('matrixChar() with supplied lines', () => {
    const lines = ['0123456789', 'ABCDEFGHIK', 'abcdefghijk']
    const i = new InputData({ lines })
    expect(i.matrixChar(1, 3)).toBe('D')
})

test('subMatrix()', () => {
    const i = new InputData({ filename: 'input-matrix.txt' })
    expect(i.subMatrix(1, 1, 2, 2)[0]).toBe('.#')
    expect(i.subMatrix(1, 1, 2, 2)[1]).toBe('..')
})